Additional docs on developing the application (#659)
apollographql/apollo-mcp-server: docs/source/mcp-apps-development.mdx

---

> title: Developing MCP Apps

Author: I changed my mind about this name. I don't really see this as "advanced" development (which makes it sound like you're using features that you won't need until complex situations). Instead its more of a "here are the tenents of developing an MCP app with this package". 

A great example of this is the [Suspense documentation](https://www.apollographql.com/docs/react/data/suspense) for Apollo Client. This isn't really "advanced" development, but rather teaching you how to use the various aspects of the suspense hooks so that you can write them yourself. This is the style of doc I'm going for here. I just wanted a place where we have the various aspects of developing the application in one place.

Also to note, the "Reference" section I see as more of the comprehensive "what options are available in x or y API", so its less about teaching those APIs, and more about giving you whats available.

With that in mind, do you have any other ideas on what we might title this page?

> title: Developing MCP Apps

Author: I'd be ok with one of the first two options. Let's use that for now. We can always switch later. Thanks!

subtitle: Key concepts when developing MCP Apps
---

This reference guide provides an overview of the core concepts of developing an MCP app using React, Apollo Client, and Apollo MCP Server.

> If you are using the [Apollo AI Apps Template](https://github.com/apollographql/ai-apps-template) to follow along, much of the following code is provided for you. This guide walks through the critical areas of developing an MCP application for your own understanding.

Reviewer: ```suggestion
Use the [Apollo AI Apps Template](https://github.com/apollographql/ai-apps-template) to get started developing your app. The template includes much of the code referenced in this document.
```

Author: This may or may not be true by the time I'm done with this doc. Let me revisit this when I'm done 😄

Author: @pragl with the latest pushed updates, I'd love for you to read the document as a whole and see if you think we should include this statement. Here are my suggestions:

I would recommend we use "if", not just "use" here as my thinking is that it should be for anyone that wants to write code for an MCP app. The AI template provides some structure up-front, but it is not required to go through this document. That said, if you think we should urge that as the starting point always, then I agree we should change it to be a more definitive statement.

I think we should probably include the 2nd sentence the way you wrote it. I'll wait to update though until we finalize the first sentence so that this one makes sense 🙂

<Note>

This guide does not provide a comprehensive overview of Apollo Client capabilities. Refer to the [Apollo Client documentation](../react) for more documentation about Apollo Client.

</Note>

## Get started

The `@apollo/client-ai-apps` package uses Apollo Client to manage GraphQL operations in our MCP app. This means you can use the knowledge you already have building applications with Apollo Client to build MCP apps.

### Initialize Apollo Client

To get started, we need to initialize and instance of `ApolloClient` from the `@apollo/client-ai-apps` package. The `ApolloClient` constructor needs, at minim, an instance of an `ApolloCache` and our [manifest file](./mcp-apps-reference#manifest-file).

<Note>

Make sure you import `ApolloClient` from the `@apollo/client-ai-apps` package, not the core `@apollo/client` repository. The `ApolloClient` class from `@apollo/client-ai-apps` handles the details of integrating with an MCP apps environment.

> </Note>

Reviewer: Tag as important or caution rather than a note?

```ts
// src/main.tsx
import { InMemoryCache } from "@apollo/client";
import { ApolloClient, ApplicationManifest } from "@apollo/client-ai-apps";
// Note the manifest location is at the root of our app
import manifest from "../.application-manifest.json";

// You may also provide other options that the core ApolloClient accepts
const client = new ApolloClient({
  cache: new InMemoryCache(),
  manifest: manifest as ApplicationManifest,
});
```

<Note>

> Unlike the `ApolloClient` class from the core `@apollo/client` package, the `ApolloClient` class from `@apollo/client-ai-apps` does not require a configured `link`. By default it uses `ToolCallLink` which executes GraphQL queries through MCP tools. <br/><br/>

Reviewer: ```suggestion
Unlike the `ApolloClient` class from the core `@apollo/client` package, the `ApolloClient` class from `@apollo/client-ai-apps` doesn't require a configured `link`. By default, the class uses `ToolCallLink`, which executes GraphQL queries through MCP tools. <br/><br/>
```

You may provide a custom link chain to provide additional capability to `ApolloClient`, however your terminating link must be a `ToolCallLink`.

</Note>

### Provide your client to `ApolloProvider`

Once you have a `client` instance, you need to provide that client to `ApolloProvider`. Like `ApolloClient`, you will need to use the `ApolloProvider` component exported from the `@apollo/client-ai-apps` package. This implementaton is reponsible for initializing the client with our MCP server.

```tsx
import { ApolloProvider } from "@apollo/client-ai-apps/react";

createRoot(document.getElementById("root")!).render(
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>,
);
```

## Register an operation as an MCP tool

> Tools are defined from your GraphQL operations using the `@tool` directive. You write queries like you would in an Apollo Client application and add the `@tool` directive to the operation.

Reviewer: ```suggestion
The `@tool` directive defines tools from your GraphQL operations, so you can write queries like you would in an Apollo Client application. Just add the `@tool` directive to the operation.
```

```tsx
import { gql, TypedDocumentNode } from "@apollo/client";
import { useQuery } from "@apollo/client/react";

// Note: These types should be generated using a tool like GraphQL Codegen.
// Avoid writing them by hand.
const TOP_PRODUCTS_QUERY: TypedDocumentNode<
  TopProductsQuery,
  TopProductsQueryVariables
> = gql`
  query TopProductsQuery
  @tool(
    name: "TopProducts"
    description: "Shows a list of the highest rated products."
  ) {
    topProducts {
      id
      title
      rating
      price
      thumbnail
    }
  }
`;

function TopProducts() {
  const { data, loading, error, dataState } = useQuery(TOP_PRODUCTS_QUERY);

  if (loading) {
    return <div>Loading...</div>;
  }

  if (error) {
    return <div>Error! {error.message}</div>;
  }

  return (
    <div>
      {data.topProducts.map((product) => (
        <Product key={product.id} product={product} />
      ))}
    </div>
  );
}
```

> Now this query is registered as a tool in our app! If you run the app in ChatGPT or an MCP apps compatible host, you can ask the LLM to show you some top products. This will render our view using data fetched by our MCP server.

Reviewer: ```suggestion
Now that the query is registered as a tool, you can run the app in ChatGPT or an MCP Apps-compatible host and prompt the LLM to show you the top products. They render using data fetched by MCP Server.
```

<Caution>

Make sure you import `ApolloClient` from the `@apollo/client-ai-apps` package, not the core `@apollo/client` repository. The `ApolloClient` class from `@apollo/client-ai-apps` is for integrating with the MCP Apps environment.

</Caution>

```ts title=src/main.tsx
import { InMemoryCache } from "@apollo/client";
import { ApolloClient, ApplicationManifest } from "@apollo/client-ai-apps";
// Note the manifest is written to the root of the app
import manifest from "../.application-manifest.json";

// You may also provide other options that the core ApolloClient accepts
const client = new ApolloClient({
  cache: new InMemoryCache(),
  manifest: manifest as ApplicationManifest,
});
```

<Note>

Unlike the `ApolloClient` class from the core `@apollo/client` package, the `ApolloClient` class from `@apollo/client-ai-apps` doesn't require a configured `link`. By default, the class uses `ToolCallLink`, which executes GraphQL queries through MCP tools. <br/><br/>

You can provide a custom link chain to grant additional capabilities to `ApolloClient`; however, your terminating link needs to be a `ToolCallLink`.

</Note>

### Provide your client to `ApolloProvider`

> Once you've created your `client` instance, provide that client to `ApolloProvider`. Like `ApolloClient`, you need to use the `ApolloProvider` component exported from the `@apollo/client-ai-apps` package.

Reviewer: That works!

```tsx title=src/main.tsx
import { ApolloProvider } from "@apollo/client-ai-apps/react";

const client = new ApolloClient({
  // ...
});

createRoot(document.getElementById("root")!).render(
  <ApolloProvider client={client}>
    <App />
  </ApolloProvider>,
);
```

> The `ApolloProvider` implementation from `@apollo/client-ai-apps/react` is responsible for initializing your application with the MCP server and providing the `client` instance in React context for use with Apollo Client's data fetching hooks.

Reviewer: ```suggestion
The `ApolloProvider` implementation from `@apollo/client-ai-apps/react` initializes your application with MCP Server and provides the `client` instance in React context for use with Apollo Client's data-fetching hooks.
```

#### Show a loading fallback during initialization

The `ApolloProvider` component leverages React's [Suspense](https://react.dev/reference/react/Suspense) functionality during initialization to avoid rendering application code that relies on client initialization. The client is initialized once the `ui/notifications/tool-result` notification is received from the host.

> You can use React's [`Suspense` component](https://react.dev/reference/react/Suspense) to show a loading fallback while the app is initializing. Wrap `ApolloProvider` with `Suspense`.

Reviewer: ```suggestion
If you want to use React's [`Suspense` component](https://react.dev/reference/react/Suspense) to display a loading fallback while the app initializes, wrap `ApolloProvider` with `Suspense`.
```

Reviewer: Yes! This works great.

import { useWidgetState } from "@apollo/client-ai-apps/openai";

export function MyComponent() {
  const [widgetState, setWidgetState] = useWidgetState();

  return (
    <div>
      <div>{widgetState.foo.toString()}</div>
      <button onClick={() => setWidgetState({ foo: true })}>Set foo</button>
    </div>
  );
}
```

<Tip>

For additional type safety when developing a platform-specific module, please reference the [TypeScript configuration](#typescript-configuration) section to update your configuration.

</Tip>

### Import a platform-specific module

You import these files using the base name. For example, if you create a platform-specific `Home` component for OpenAI (`Home.openai.tsx`) and MCP (`Home.mcp.tsx`), you would import it as follows:

```tsx
import { Home } from "./Home";
```

<Caution>

You need to make sure that each configured environment includes an importable module, otherwise the application will fail to build. For example, if you create a `Home.openai.tsx` file, you will also need to create either a `Home.mcp.tsx` or `Home.tsx` file to ensure that the MCP apps environment builds correctly.

</Caution>

## TypeScript configuration

> `@apollo/client-ai-apps` provides extendable TypeScript configs that ensures you only use code available to your application that might be available in [platform-specific modules](#platform-specific-modules).

Reviewer: ```suggestion
`@apollo/client-ai-apps` provides extendable TypeScript configurations that make sure you only use code available to your application from [platform-specific modules](#platform-specific-modules).
```

- `@apollo/client-ai-apps/tsconfig/core` — for shared modules that run in both environments. This config ensures that included files only use shared utilities available to all environments. Use this when type-checking shared files used in all environments.
- `@apollo/client-ai-apps/tsconfig/mcp` — for developing MCP-specific modules. Allows you to use utilities exported by `@apollo/client-ai-apps/mcp`. Use this when type-checking your `.mcp.*` files.

> - `@apollo/client-ai-apps/tsconfig/openai` — for developing ChatGPT-specific modules. Allows you to use utilities exported by `@apollo/client-ai-apps/openai`. Use this when type-checking your `.openai.*` files.

Reviewer: ```suggestion
- `@apollo/client-ai-apps/tsconfig/openai` — For developing ChatGPT-specific modules. This configuration allows you to use utilities exported by `@apollo/client-ai-apps/openai`. Use this when type-checking your `.openai.*` files.
```

### Recommended TypeScript configuration

Use TypeScript [project references](https://www.typescriptlang.org/docs/handbook/project-references.html) to apply a TypeScript configuration to a subset of files. We recommend the following config files:

```
.
├── tsconfig.app.json
├── tsconfig.base.json
├── tsconfig.mcp.json
├── tsconfig.openai.json
└── tsconfig.json
```

**`tsconfig.base.json`**

> Shared configuration for all TypeScript files. This should be where most of your `compilerOptions` are configured.

Reviewer: ```suggestion
Shared configuration for all TypeScript files. This should be where you should configure most of your `compilerOptions`.
```

Author: Could we remove the double "should" in the 2nd sentence?

```suggestion
Shared configuration for all TypeScript files. This is where you should configure most of your `compilerOptions`.
```

  "include": ["src"],
  "exclude": ["src/**/*.mcp.*", "src/**/*.openai.*"],
}
```

**`tsconfig.mcp.json`**

Configuration that targets MCP app specific modules. Extend `@apollo/client-ai-apps/tsconfig/mcp` and include `.mcp.*` files.

```jsonc title=tsconfig.mcp.json
{
  "extends": ["@apollo/client-ai-apps/tsconfig/mcp", "./tsconfig.base.json"],
  "include": ["src/**/*.mcp.ts", "src/**/*.mcp.tsx"],
}
```

<Note>

If your app doesn't have any MCP apps specific files, don't include this file because TypeScript might complain that it is unable to match files in the `include` setting.

</Note>

**`tsconfig.openai.json`**

Configuration that targets ChatGPT app specific modules. Extend `@apollo/client-ai-apps/tsconfig/openai` and include `.openai.*` files. Include the `openai/globals` types in `types` configuration which provides types for `window.openai`.

```jsonc title=tsconfig.openai.json
{
  "extends": ["@apollo/client-ai-apps/tsconfig/openai", "./tsconfig.base.json"],
  "compilerOptions": {
    "types": ["@apollo/client-ai-apps/openai/globals"],
  },
  "include": ["src/**/*.openai.ts", "src/**/*.openai.tsx"],
}
```

<Note>

> If your app doesn't have any MCP apps specific files, don't include this file because TypeScript might complain that it is unable to match files in the `include` setting.

Reviewer: ```suggestion
If your app doesn't have any MCP Apps-specific files, don't include this file because TypeScript might be unable to match files in the `include` setting.
```

</Note>

**`tsconfig.json`**

The base config that should include all project references

```jsonc title=tsconfig.openai.json
{
  "files": [],
  "references": [
    { "path": "./tsconfig.app.json" },
    // Omit if you don't create this file
    { "path": "./tsconfig.mcp.json" },
    // Omit if you don't create this file
    { "path": "./tsconfig.openai.json" },
  ],
}
```

## Vite configuration

The `@apollo/client-ai-apps` package includes a [Vite](https://vite.dev/) plugin that configures your application for developing ChatGPT and MCP apps.

> To use it, import `apolloClientAiApps` from `@apollo/client-ai-apps/vite` and provide it to the `plugins` option in your vite config. Set the `targets` option to tell the plugin which environments you are building for.

Reviewer: ```suggestion
To use it, import `apolloClientAiApps` from `@apollo/client-ai-apps/vite` and provide `apolloClientAiApps` to the `plugins` option in your Vite configuration. Set the `targets` option to define what environments you are building for.
```

- `mcp` - Creates a build artifact for MCP apps

> - `openai` - Creates a build artifact for ChatGPT apps

Reviewer: ```suggestion
- `openai` - Creates a build artifact for ChatGPT apps.
```

```ts title=vite.config.ts
import { defineConfig } from "vite";
import { apolloClientAiApps } from "@apollo/client-ai-apps/vite";

export default defineConfig({
  plugins: [
    apolloClientAiApps({
      targets: ["mcp", "openai"],
    }),
    // ... other plugins
  ],
  // ... other config
});
```

### Development mode

You can only run a single environment in development mode at a time. Set the `devTarget` option to tell the plugin which environment you are developing for.

```ts
apolloClientAiApps({
  devTarget: "openai",
  // ...
});
```

<Note>

> If the configured `targets` option only includes a single value, you may omit the `devTarget` option because the plugin will infer the `devTarget` as the value in `targets`.

Reviewer: ```suggestion
If the configured `targets` option only includes one value, you can omit the `devTarget` option because the plugin infers the `devTarget` as the value in `targets`.
```
